### PR TITLE
Base Calculations [Epic #142][User Stories #143 #144][Tasks #146 #147 #148 # 149 # 150]

### PR DESCRIPTION
**Pull request by todd829**
Implements User Stories #143 #144

### Description
Add a new view for bitwise calculations involving binary, octal and decimal numbers.

### Demonstration
bitwisedemo

### WHAT kind of change does this PR introduce?

New view for bitwise calculations involving binary, octal and decimal numbers.

### HOW is this accomplished?

This is done by creating new bitwise operations in the operation package for bitwise operations such as *and &*, *or |*, *xor ^* and *inv ~*, a new view, a new activity and conversion functions.

### Checklist
- [x] Docs have been added / updated to reflect the changes
- [x] I have reviewed and tested the changes :white_check_mark:
- [x] I have squashed my commits to have a reasonable amount of commits
- [x] All of the code I have written adhere to the coding conventions outlined in the coding conventions section of the wiki

### CODE REVIEW

**Comment by jusleg** [comment]
@todd829 it doesn't work with anything above 1023.

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/BitwiseActivity.kt [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+import com.simpletools.calculator.commons.extensions.config
+import com.simpletools.calculator.commons.extensions.updateViewColors
+
+/* ktlint-disable no-wildcard-imports */
+import com.simpletools.calculator.commons.helpers.*
+import kotlinx.android.synthetic.main.activity_base.*
+import com.simplemobiletools.commons.extensions.*
+/* ktlint-enable no-wildcard-imports */
+
+class BitwiseActivity : SimpleActivity(), Calculator {
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/BitwiseActivity.kt]: Glad the name was changed clears up alot of confusion!

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/BitwiseCalculatorImpl.kt [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+                setValue(firstNumber.toString())
+                lastIsOperation = false
+            } else {
+                var value = firstNumber
+                firstNumber = convertBin(OCT, value)
+                setValue(firstNumber.toString())
+                lastIsOperation = false
+            }
+        } catch (e: NumberFormatException) {
+            setValue("OVERFLOW")
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/helpers/BitwiseCalculatorImpl.kt]: Good catch !

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/BitwiseCalculatorImpl.kt [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+        setFormula("")
+        digits = 0
+    }
+
+    fun addDigit(i: Long) {
+        if (operator != "") {
+            secondNumberSet = true
+            setClear()
+        }
+
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/helpers/BitwiseCalculatorImpl.kt]: I am not 100% sure but I believe that's one too many spaces in here.

File: commons/src/main/java/com/simpletools/calculator/commons/operations/InvOperation.kt [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package com.simpletools.calculator.commons.operations
+
+import com.simpletools.calculator.commons.operations.base.UnaryBitwiseOperation
+
+object InvOperation : UnaryBitwiseOperation() {
+
+    override fun getResult() = value.inv()
+
+    override fun applyOperator(str: String): String {
+        return "~($str)"
```
Comment (almiche) [inline commons/src/main/java/com/simpletools/calculator/commons/operations/InvOperation.kt]: What does the ~ signify ?

**Comment by almiche** [comment]
❌Styling is adequate and follows coding conventions outlined by the [Kotlin Foundation](https://kotlinlang.org/docs/reference/coding-conventions.html)

* Certain places have a few too many spaces between lines of code it should always be one line between overall very good

✅ New methods and functionalities have appropriate tests (testing edge cases + good coverage)

✅ Branch has an acceptable amount of commits with descriptive names and

* A few too many commits but the names are descriptive and I understand the reasoning behind why it was done this way

✅   PR description matches the code outlined a has mentions to issues

✅ This PR attempts to implement a specific feature (no code blob or epic pr)

Status: **Changes needed** Good PR overall just waiting on some answers to my questions and will approve

File: commons/src/main/java/com/simpletools/calculator/commons/operations/InvOperation.kt (hunk 4 above)
Comment (todd829) [reply commons/src/main/java/com/simpletools/calculator/commons/operations/InvOperation.kt]: inv operation, so formula for the inv of 5 is ~(5)

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/BitwiseCalculatorImpl.kt (hunk 3 above)
Comment (todd829) [reply app/src/main/kotlin/com/simplemobiletools/calculator/helpers/BitwiseCalculatorImpl.kt]: Both ways are correct I believe in that case. I like to have my 'if' operations surrounded by single space unless the 'if' is followed by nothing but a closing bracket like a method ending or an 'if' inside an 'if', you will see:
    }
}

And never: 
    }

}

In that particular case, operations are done after the 'if' statement and I find it cleaner to add a space in between. The choice is yours, I can make a new commit for this single space!

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/BitwiseActivity.kt (hunk 1 above)
Comment (todd829) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/BitwiseActivity.kt]: Indeed

**Comment by almiche** [comment]
✅Styling is adequate and follows coding conventions outlined by the [Kotlin Foundation](https://kotlinlang.org/docs/reference/coding-conventions.html)

✅ New methods and functionalities have appropriate tests (testing edge cases + good coverage)

✅ Branch has an acceptable amount of commits with descriptive names and

✅   PR description matches the code outlined a has mentions to issues

✅ This PR attempts to implement a specific feature (no code blob or epic pr)

Status: **Approved** Good PR overall
